File: app/context/memory.py
"""
Memory management MCP server.
"""

from typing import Annotated, Any

from uuid import UUID
from fastmcp import FastMCP
from fastmcp.server.dependencies import get_http_headers

from app.util.ingest import INGEST_AVAILABLE, ingest_document
from app.util.memory import (
    add_memory,
    buckets_list,
    clear_memories,
    connect_memories,
    delete_memory,
    disconnect_memories,
    get_connected,
    get_last_memories,
    get_memory,
)

__all__ = ["server"]

server = FastMCP("Memory Tools")


def _get_namespace() -> str:
    """Extract namespace from headers or use default."""
    headers = get_http_headers()
    return headers.get("x-namespace", "default")


@server.tool(tags={"memory"})
async def add(
    content: Annotated[str, "Content to store in memory"],
    bucket: Annotated[str | None, "Optional bucket name"] = None,
    connected_nodes: Annotated[list[str] | None, "Optional list of memory UUIDs this node connects to"] = None,
    relationship_types: Annotated[
        list[str] | None, "Parallel list of relationship labels (same length as connected_nodes)"
    ] = None,
) -> dict[str, str]:
    """
    Add content to memory storage, optionally with initial graph edges.

    Args:
        content: Content to store in memory
        bucket: Optional bucket name
        connected_nodes: Optional list of memory UUID strings to connect to
        relationship_types: Parallel list of relationship labels

    Returns:
        dict[str, str]: The UUID of the stored memory item.
    """
    namespace = _get_namespace()
    memory_id = await add_memory(content, bucket, namespace, connected_nodes, relationship_types)
    return {
        "status": "Memory added",
        "memory_id": str(memory_id),
    }


@server.tool(tags={"memory"})
async def get(
    query: Annotated[str, "Query to search in memory"],
    bucket: Annotated[str | None, "Optional bucket name"] = None,
    top_k: Annotated[int, "Number of top results to return"] = 5,
) -> dict[str, Any] | None:
    """
    Retrieve content from memory storage based on a query.
    Args:
        query: Query to search in memory
        bucket: Optional bucket name
        top_k: Number of top results to return
    Returns:
        dict[str, Any] | None: The retrieved memory item or None if not found.
    """
    namespace = _get_namespace()
    results = await get_memory(query, bucket, namespace, top_k=top_k)

    return {
        "status": "Memory retrieved" if results else "No memory found",
        "results": results,
    }


@server.tool(tags={"memory"})
async def get_last(
    n: Annotated[int, "Number of recent memories to return"] = 5,
    bucket: Annotated[str | None, "Optional bucket name"] = None,
) -> dict[str, Any]:
    """
    Retrieve the last N memories ordered by creation time (most recent first).

    Args:
        n: Number of recent memories to return (default 5)
        bucket: Optional bucket name to filter by

    Returns:
        dict[str, Any]: The retrieved memory items.
    """
    namespace = _get_namespace()
    results, _ = await get_last_memories(n, bucket, namespace)

    return {
        "status": "Memory retrieved" if results else "No memory found",
        "results": results,
    }


@server.tool(tags={"memory"})
async def delete(
    memory_id: Annotated[str, "UUID of the memory item to delete"],
) -> dict[str, str]:
    """
    Delete a memory item by its UUID.

    Args:
        memory_id: UUID of the memory item to delete

    Returns:
        dict[str, str]: Status message.
    """
    namespace = _get_namespace()
    await delete_memory(UUID(memory_id), namespace)
    return {
        "status": "Memory deleted",
    }


@server.tool(tags={"memory"})
async def clear(
    bucket: Annotated[str | None, "Optional bucket name to clear"] = None,
) -> dict[str, str]:
    """
    Clear all memory items, optionally within a specific bucket.

    Args:
        bucket: Optional bucket name to clear
    Returns:
        dict[str, str]: Status message.
    """
    namespace = _get_namespace()
    await clear_memories(bucket, namespace)
    return {
        "status": "Memories cleared",
    }


@server.tool(tags={"memory"})
async def list_buckets() -> dict[str, list[str]]:
    """
    List all memory buckets.

    Returns:
        dict[str, list[str]]: List of bucket names.
    """
    namespace = _get_namespace()
    buckets = await buckets_list(namespace)
    return {
        "buckets": list(buckets),
    }


# ---- Knowledge-graph tools ----


@server.tool(tags={"memory", "graph"})
async def connect(
    source_id: Annotated[str, "UUID of the source memory node"],
    target_id: Annotated[str, "UUID of the target memory node"],
    relationship_type: Annotated[str, "Label for the directed edge (e.g. 'related_to', 'depends_on')"],
) -> dict[str, str]:
    """
    Create a directed edge between two memory nodes.

    Args:
        source_id: UUID of the source memory
        target_id: UUID of the target memory
        relationship_type: Label describing the relationship

    Returns:
        dict[str, str]: Status message.
    """
    namespace = _get_namespace()
    await connect_memories(UUID(source_id), UUID(target_id), relationship_type, namespace)
    return {"status": "Memories connected"}


@server.tool(tags={"memory", "graph"})
async def disconnect(
    source_id: Annotated[str, "UUID of the source memory node"],
    target_id: Annotated[str, "UUID of the target memory node"],
    relationship_type: Annotated[str | None, "Remove only this edge label; omit to remove all edges"] = None,
) -> dict[str, str]:
    """
    Remove one or all directed edges between two memory nodes.

    Args:
        source_id: UUID of the source memory
        target_id: UUID of the target memory
        relationship_type: If given, only remove the matching edge

    Returns:
        dict[str, str]: Status message.
    """
    namespace = _get_namespace()
    await disconnect_memories(UUID(source_id), UUID(target_id), relationship_type, namespace)
    return {"status": "Memories disconnected"}


@server.tool(tags={"memory", "graph"})
async def traverse(
    memory_id: Annotated[str, "UUID of the starting memory node"],
    relationship_type: Annotated[str | None, "Filter traversal to this relationship type"] = None,
    depth: Annotated[int, "How many hops to traverse (default 1)"] = 1,
) -> dict[str, Any]:
    """
    Traverse the knowledge graph starting from a memory node.

    Returns connected memories up to *depth* hops away, optionally filtered
    by relationship type.

    Args:
        memory_id: UUID of the starting memory
        relationship_type: Optional edge-label filter
        depth: Number of hops (1 = direct neighbours only)

    Returns:
        dict: Status and list of connected memory nodes with depth info.
    """
    namespace = _get_namespace()
    results = await get_connected(UUID(memory_id), namespace, relationship_type, depth)
    return {
        "status": "Graph traversed" if results else "No connected nodes found",
        "results": results,
    }


# ---- Document ingestion (optional add-on) ----


if INGEST_AVAILABLE:

    @server.tool(tags={"memory", "ingest"})
    async def ingest(
        text: Annotated[str, "Document text to chunk and store"],
        source: Annotated[str, "Source name; seeds the bucket name"],
        bucket: Annotated[str | None, "Override bucket (defaults to a name derived from source)"] = None,
        replace: Annotated[bool, "Clear the target bucket before ingesting"] = False,
        parent: Annotated[bool, "Create a parent node the chunks attach to"] = True,
        parent_content: Annotated[str | None, "Content for the parent node (defaults to source)"] = None,
    ) -> dict[str, Any]:
        """
        Chunk a document and store the chunks as memories.

        Splits *text* into semantically coherent chunks and stores each as a memory in a
        per-document bucket (derived from *source* unless *bucket* is given). Set *replace*
        to clear the bucket first for idempotent re-ingestion. When *parent* is set a node
        (content *parent_content*, defaulting to *source*) anchors the chunks via ``part_of``.

        Args:
            text: Document text to chunk and store
            source: Source name; seeds the bucket name
            bucket: Optional explicit bucket name
            replace: Clear the target bucket before ingesting
            parent: Create a parent node the chunks attach to
            parent_content: Content for the parent node (defaults to source)

        Returns:
            dict: status, bucket, chunk count, the stored memory UUIDs, and the parent id.
        """
        namespace = _get_namespace()
        result = await ingest_document(
            text,
            name=source,
            bucket=bucket,
            namespace=namespace,
            replace=replace,
            parent=parent,
            parent_content=parent_content,
        )
        return {
            "status": "Document unchanged" if result["skipped"] else "Document ingested",
            "bucket": result["bucket"],
            "chunks": result["chunks"],
            "memory_ids": [str(m) for m in result["memory_ids"]],
            "skipped": result["skipped"],
            "parent_id": result["parent_id"],
        }

File: app/static/js/app/memory.js
document.addEventListener("DOMContentLoaded", () => {
  const { createApp, ref, computed, onMounted, nextTick } = Vue;

  createApp({
    setup() {
      // Shared auth / namespace / bucket / toast state and API helper.
      const shared = ArcaShared.create(Vue);
      const {
        authKey, authKeyInput, namespace, namespaces, connected, buckets,
        error, success, flash, showError, api, loadNamespaces, loadBuckets,
        saveAuth, disconnectAuth, showNsForm, newNsName, createNamespace, canvasPosKey,
      } = shared;

      // ── Data ───────────────────────────────────────────────
      const activeBucket = ref(null);
      const memories = ref([]);
      const isSearchResult = ref(false);
      const searchQuery = ref("");

      // ── Pagination (recent memories list) ──────────────────
      const page = ref(0);
      const pageSize = ref(50);
      const total = ref(0);
      const pageStart = computed(() => (total.value === 0 ? 0 : page.value * pageSize.value + 1));
      const pageEnd = computed(() => Math.min((page.value + 1) * pageSize.value, total.value));
      const hasPrevPage = computed(() => page.value > 0);
      const hasNextPage = computed(() => (page.value + 1) * pageSize.value < total.value);

      // ── Add memory ─────────────────────────────────────────
      const showAddForm = ref(false);
      const newMemory = ref({ content: "", bucket: "", newBucket: "" });
      const saving = ref(false);

      // ── Ingest document ────────────────────────────────────
      const showIngestForm = ref(false);
      const ingestFiles = ref([]);         // one or more selected files, ingested into one bucket
      const ingestBucket = ref("");        // "" = derive from file name, a bucket name, or "__new__"
      const ingestNewBucket = ref("");     // name typed when ingestBucket === "__new__"
      const ingestReplace = ref(false);
      const ingestCreateParent = ref(false);  // opt-in: anchor the chunks to a parent node
      const ingestParentContent = ref("");    // parent node content (blank → file name / bucket name)
      const ingesting = ref(false);
      const ingestProgress = ref("");      // "3 / 10" while a multi-file batch uploads
      const ingestDragging = ref(false);
      const ingestAvailable = ref(true);   // optimistic until /ingest/formats answers
      const supportedFormats = ref([]);    // file extensions the server can actually parse
      const acceptAttr = computed(() => supportedFormats.value.join(","));
      const formatsLabel = computed(() =>
        supportedFormats.value.length ? supportedFormats.value.join(", ") : ".txt, .md"
      );
      const ingestTotalKb = computed(() =>
        (ingestFiles.value.reduce((s, f) => s + f.size, 0) / 1024).toFixed(1)
      );
      // Resolved target bucket: "" means "let the server derive it from the file name".
      const ingestEffectiveBucket = computed(() =>
        ingestBucket.value === "__new__" ? ingestNewBucket.value.trim() : ingestBucket.value
      );
      // A target bucket is required when ingesting more than one file (no single filename to derive from).
      const canIngest = computed(() =>
        ingestFiles.value.length > 0 && !ingesting.value &&
        !(ingestBucket.value === "__new__" && !ingestNewBucket.value.trim()) &&
        !(ingestFiles.value.length > 1 && !ingestEffectiveBucket.value)
      );

      // ── Inline edit ────────────────────────────────────────
      const editingId = ref(null);
      const editContent = ref("");
      const editBucket = ref("");
      const editNewBucket = ref("");
      const updatingId = ref(null);

      // ── Connections ────────────────────────────────────────
      const expandedId = ref(null);
      const newEdge = ref({ query: "", targetId: "", relType: "" });
      const edgeResults = ref([]);
      const edgeSearching = ref(false);

      // ── Bucket management ──────────────────────────────────
      const showBucketForm = ref(false);
      const newBucketName = ref("");
      const renamingBucket = ref(null);
      const renameBucketValue = ref("");

      // ── UI state ───────────────────────────────────────────
      const loading = ref(false);
      const searching = ref(false);

      // ── Buckets ────────────────────────────────────────────

      // Buckets only materialize server-side when their first memory lands, so
      // UI-created names are tracked here and merged into every bucket refresh —
      // otherwise they'd silently vanish on the next reload.
      const pendingBuckets = new Set();

      function mergePendingBuckets() {
        for (const name of [...pendingBuckets]) {
          if (buckets.value.includes(name)) pendingBuckets.delete(name);
          else buckets.value.push(name);
        }
        if (pendingBuckets.size) buckets.value.sort();
      }

      async function refreshBuckets() {
        await loadBuckets();
        mergePendingBuckets();
      }

      async function createBucket() {
        const name = newBucketName.value.trim();
        if (!name) return;
        if (buckets.value.includes(name)) {
          showError("Bucket already exists");
          return;
        }
        pendingBuckets.add(name);
        buckets.value.push(name);
        buckets.value.sort();
        newBucketName.value = "";
        showBucketForm.value = false;
        flash(`Bucket "${name}" ready — it persists once a memory is added`);
      }

      function startRenameBucket(name) {
        renamingBucket.value = name;
        renameBucketValue.value = name;
      }

      async function confirmRenameBucket(oldName) {
        const newName = renameBucketValue.value.trim();
        if (!newName || newName === oldName) {
          renamingBucket.value = null;
          return;
        }
        try {
          await api("POST", "/buckets/rename", { old_name: oldName, new_name: newName });
          renamingBucket.value = null;
          flash(`Bucket renamed to "${newName}"`);
          if (activeBucket.value === oldName) activeBucket.value = newName;
          // Carry the canvas layout over to the renamed bucket.
          const savedLayout = localStorage.getItem(canvasPosKey(oldName));
          if (savedLayout) {
            localStorage.setItem(canvasPosKey(newName), savedLayout);
            localStorage.removeItem(canvasPosKey(oldName));
          }
          await reload();
        } catch (e) {
          showError("Rename failed: " + e.message);
        }
      }

      async function clearBucket(name) {
        const ok = await window.ArcaConfirm.open({
          title: "Clear bucket",
          message: `This will permanently delete ALL memories in "${name}". This cannot be undone.`,
          confirmLabel: "Clear all",
          tone: "danger",
        });
        if (!ok) return;
        try {
          await api("DELETE", `/memories?bucket=${encodeURIComponent(name)}`);
          flash(`Bucket "${name}" cleared`);
          if (activeBucket.value === name) activeBucket.value = null;
          localStorage.removeItem(canvasPosKey(name));
          await reload();
        } catch (e) {
          showError("Clear failed: " + e.message);
        }
      }

      function selectBucket(b) {
        activeBucket.value = b;
        page.value = 0;
        if (isSearchResult.value) {
          searchMemories();
        } else {
          loadMemories();
        }
      }

      // ── Memories ───────────────────────────────────────────

      async function loadMemories() {
        loading.value = true;
        isSearchResult.value = false;
        try {
          const body = { n: pageSize.value, offset: page.value * pageSize.value };
          if (activeBucket.value) body.bucket = activeBucket.value;
          const data = await api("POST", "/memories/last", body);
          memories.value = data.results || [];
          total.value = data.total ?? memories.value.length;
        } catch (e) {
          showError("Failed to load memories: " + e.message);
        } finally {
          loading.value = false;
        }
      }

      function nextPage() {
        if (!hasNextPage.value) return;
        page.value++;
        loadMemories();
      }

      function prevPage() {
        if (!hasPrevPage.value) return;
        page.value--;
        loadMemories();
      }

      function changePageSize() {
        page.value = 0;
        loadMemories();
      }

      async function searchMemories() {
        if (!searchQuery.value.trim()) return;
        searching.value = true;
        try {
          const body = { query: searchQuery.value, top_k: 20 };
          if (activeBucket.value) body.bucket = activeBucket.value;
          const data = await api("POST", "/memories/search", body);
          memories.value = data.results || [];
          isSearchResult.value = true;
        } catch (e) {
          showError("Search failed: " + e.message);
        } finally {
          searching.value = false;
        }
      }

      function clearSearch() {
        searchQuery.value = "";
        isSearchResult.value = false;
        page.value = 0;
        loadMemories();
      }

      async function addMemory() {
        if (!newMemory.value.content.trim()) return;
        saving.value = true;
        try {
          const body = { content: newMemory.value.content };
          // Dropdown value is a bucket name, "" (default bucket), or "__new__" (use the typed name).
          const bucket = newMemory.value.bucket === "__new__"
            ? newMemory.value.newBucket.trim()
            : newMemory.value.bucket.trim();
          if (bucket) body.bucket = bucket;
          await api("POST", "/memories", body);
          newMemory.value = { content: "", bucket: "", newBucket: "" };
          showAddForm.value = false;
          flash("Memory saved");
          await reload();
        } catch (e) {
          showError("Failed to save: " + e.message);
        } finally {
          saving.value = false;
        }
      }

      // ── Ingest document ────────────────────────────────────

      async function loadIngestFormats() {
        try {
          const data = await api("GET", "/ingest/formats");
          ingestAvailable.value = data.available;
          supportedFormats.value = data.extensions || [];
        } catch {
          // Non-fatal: keep optimistic defaults so ingestion still works if the probe fails.
        }
      }

      function extOf(name) {
        const i = name.lastIndexOf(".");
        return i === -1 ? "" : name.slice(i).toLowerCase();
      }

      function setIngestFiles(fileList) {
        const incoming = Array.from(fileList || []);
        if (!incoming.length) return;
        const exts = supportedFormats.value;
        const accepted = [];
        const rejected = [];
        for (const f of incoming) {
          // Drag-and-drop bypasses the file picker's `accept`, so validate here too.
          if (exts.length && !exts.includes(extOf(f.name))) rejected.push(f.name);
          else accepted.push(f);
        }
        if (rejected.length) {
          showError(`Unsupported file type(s): ${rejected.join(", ")}. Supported: ${exts.join(", ")}`);
        }
        if (accepted.length) ingestFiles.value = accepted;
      }

      function onIngestPick(e) {
        setIngestFiles(e.target.files);
        e.target.value = "";  // allow re-picking the same file(s) later
      }

      function onIngestDrop(e) {
        ingestDragging.value = false;
        setIngestFiles(e.dataTransfer.files);
      }

      async function ingestDocument() {
        if (!ingestFiles.value.length) return;
        const files = ingestFiles.value.slice();
        const bucketOverride = ingestEffectiveBucket.value;
        if (files.length > 1 && !bucketOverride) {
          showError("Choose a target bucket for multiple files");
          return;
        }
        const createParent = ingestCreateParent.value;
        // Parent content defaults to the bucket name for a batch (no single filename to use);
        // for a lone file a blank content lets the server fall back to the file name.
        const parentContent = ingestParentContent.value.trim() || (files.length > 1 ? bucketOverride : "");

        ingesting.value = true;
        ingestProgress.value = "";
        let okCount = 0, skipCount = 0, chunkTotal = 0;
        const failures = [];
        let targetBucket = bucketOverride || null;
        let sharedParentId = null;  // first file creates the parent; the rest attach to it
        try {
          for (let i = 0; i < files.length; i++) {
            ingestProgress.value = files.length > 1 ? `${i + 1} / ${files.length}` : "";
            const form = new FormData();
            form.append("file", files[i]);
            if (bucketOverride) form.append("bucket", bucketOverride);
            // "Replace existing" clears the whole bucket; for a batch do it once (on the first
            // file) so later files append rather than wiping their predecessors.
            form.append("replace", String(ingestReplace.value && i === 0));
            if (!createParent) {
              form.append("parent", "false");
            } else if (sharedParentId) {
              form.append("parent_id", sharedParentId);  // share one parent across the batch
            } else if (parentContent) {
              form.append("parent_content", parentContent);
            }
            try {
              // Multipart upload: let the browser set Content-Type (with boundary), so we
              // can't reuse the JSON `api()` helper.
              const res = await fetch("/v1/ingest", {
                method: "POST",
                headers: {
                  "Authorization": `Bearer ${authKey.value}`,
                  "X-Namespace": namespace.value || "default",
                },
                body: form,
              });
              if (!res.ok) {
                const detail = await res.text();
                throw new Error(`${res.status}: ${detail}`);
              }
              const data = await res.json();
              targetBucket = data.bucket;  // server-sanitized name (single file: derived from filename)
              chunkTotal += data.chunks;
              if (data.parent_id && !sharedParentId) sharedParentId = data.parent_id;
              if (data.skipped) skipCount++; else okCount++;
            } catch (e) {
              failures.push(`${files[i].name}: ${e.message}`);
            }
          }
        } finally {
          ingesting.value = false;
          ingestProgress.value = "";
        }

        // Reset the form before reporting / navigating.
        showIngestForm.value = false;
        ingestFiles.value = [];
        ingestBucket.value = "";
        ingestNewBucket.value = "";
        ingestReplace.value = false;
        ingestCreateParent.value = false;
        ingestParentContent.value = "";

        if (failures.length) {
          showError(`Ingest failed for ${failures.length} file(s): ${failures.join("; ")}`);
        }
        if (okCount || skipCount) {
          const parts = [];
          if (okCount) parts.push(`${okCount} ingested (${chunkTotal} chunk${chunkTotal === 1 ? "" : "s"})`);
          if (skipCount) parts.push(`${skipCount} unchanged`);
          flash(`${parts.join(", ")}${targetBucket ? ` → "${targetBucket}"` : ""}`);
          // Post-success refresh; loadBuckets / loadMemories handle their own errors so a
          // hiccup here never mislabels a successful ingest as a failure.
          await refreshBuckets();
          isSearchResult.value = false;
          if (targetBucket) selectBucket(targetBucket);
        }
      }

      async function deleteMemory(m) {
        const ok = await window.ArcaConfirm.open({
          title: "Delete memory",
          message: `Delete "${m.content.slice(0, 80)}${m.content.length > 80 ? "..." : ""}"? Connections will be cleaned up automatically.`,
          confirmLabel: "Delete",
          tone: "danger",
        });
        if (!ok) return;
        try {
          await api("DELETE", `/memories/${m.memory_id}`);
          memories.value = memories.value.filter(x => x.memory_id !== m.memory_id);
          if (expandedId.value === m.memory_id) expandedId.value = null;
          if (total.value > 0) total.value--;
          flash("Memory deleted");
          // If we emptied a non-first page, step back so the list isn't blank.
          if (!isSearchResult.value && !memories.value.length && page.value > 0) {
            page.value--;
            loadMemories();
          }
          refreshBuckets();
        } catch (e) {
          showError("Delete failed: " + e.message);
        }
      }

      // ── Inline edit ────────────────────────────────────────

      function startEdit(m) {
        editingId.value = m.memory_id;
        editContent.value = m.content;
        editBucket.value = m.bucket;
        editNewBucket.value = "";
        nextTick(() => {
          const el = document.querySelector(".edit-textarea");
          if (el) el.focus();
        });
      }

      function cancelEdit() {
        editingId.value = null;
        editContent.value = "";
        editBucket.value = "";
        editNewBucket.value = "";
      }

      async function saveEdit(m) {
        // Dropdown value is a bucket name or "__new__" (use the typed name).
        const bucket = editBucket.value === "__new__" ? editNewBucket.value.trim() : editBucket.value;
        const updates = {};
        if (editContent.value !== m.content) updates.content = editContent.value;
        if (bucket && bucket !== m.bucket) updates.bucket = bucket;
        if (!Object.keys(updates).length) {
          cancelEdit();
          return;
        }
        updatingId.value = m.memory_id;
        try {
          await api("PATCH", `/memories/${m.memory_id}`, updates);
          m.content = editContent.value;
          if (updates.bucket) m.bucket = updates.bucket;
          cancelEdit();
          flash("Memory updated");
          refreshBuckets();
        } catch (e) {
          showError("Update failed: " + e.message);
        } finally {
          updatingId.value = null;
        }
      }

      // ── Connections ────────────────────────────────────────

      function toggleConnections(m) {
        expandedId.value = expandedId.value === m.memory_id ? null : m.memory_id;
        newEdge.value = { query: "", targetId: "", relType: "" };
        edgeResults.value = [];
      }

      const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

      // Resolve the link target: a pasted UUID is taken as-is; anything else is a
      // semantic search across the namespace, offering candidates to pick from.
      async function searchEdgeTargets(m) {
        const q = newEdge.value.query.trim();
        if (!q) return;
        if (UUID_RE.test(q)) {
          newEdge.value.targetId = q;
          edgeResults.value = [];
          return;
        }
        edgeSearching.value = true;
        try {
          const data = await api("POST", "/memories/search", { query: q, top_k: 5 });
          edgeResults.value = (data.results || []).filter(r => r.memory_id !== m.memory_id);
          if (!edgeResults.value.length) showError("No matching memories found");
        } catch (e) {
          showError("Search failed: " + e.message);
        } finally {
          edgeSearching.value = false;
        }
      }

      function pickEdgeTarget(r) {
        newEdge.value.targetId = r.memory_id;
      }

      async function addEdge(m) {
        const targetId = newEdge.value.targetId;
        const relType = newEdge.value.relType.trim();
        if (!targetId || !relType) return;
        try {
          await api("POST", "/memories/connect", {
            source_id: m.memory_id,
            target_id: targetId,
            relationship_type: relType,
          });
          m.connected_nodes = [...(m.connected_nodes || []), targetId];
          m.relationship_types = [...(m.relationship_types || []), relType];
          newEdge.value = { query: "", targetId: "", relType: "" };
          edgeResults.value = [];
          flash("Connected");
        } catch (e) {
          showError("Connect failed: " + e.message);
        }
      }

      async function removeEdge(m, targetId, relType) {
        try {
          await api("POST", "/memories/disconnect", {
            source_id: m.memory_id,
            target_id: targetId,
            relationship_type: relType,
          });
          const idx = m.connected_nodes.findIndex(
            (n, i) => n === targetId && m.relationship_types[i] === relType
          );
          if (idx !== -1) {
            m.connected_nodes.splice(idx, 1);
            m.relationship_types.splice(idx, 1);
          }
          flash("Disconnected");
        } catch (e) {
          showError("Disconnect failed: " + e.message);
        }
      }

      // ── Lifecycle ──────────────────────────────────────────

      async function reload() {
        localStorage.setItem("arca_namespace", namespace.value || "default");
        page.value = 0;
        // Independent fetches; each handles its own errors.
        await Promise.all([loadNamespaces(), loadIngestFormats(), refreshBuckets(), loadMemories()]);
      }

      // Switching namespace invalidates the bucket filter, search scope, and
      // any not-yet-materialized bucket names (they are per-namespace).
      function onNsChange() {
        activeBucket.value = null;
        isSearchResult.value = false;
        searchQuery.value = "";
        pendingBuckets.clear();
        reload();
      }

      shared.onReload = reload;
      shared.onNamespaceChange = onNsChange;
      shared.onDisconnect = () => {
        memories.value = [];
        activeBucket.value = null;
      };

      function formatDate(dt) {
        if (!dt) return "-";
        const d = new Date(dt);
        return d.toLocaleString(undefined, {
          month: "short", day: "numeric",
          hour: "2-digit", minute: "2-digit",
        });
      }

      onMounted(() => {
        // Hook from the canvas page: /memory#ingest opens the ingest panel.
        if (location.hash === "#ingest") showIngestForm.value = true;
        if (authKey.value) {
          nextTick(() => reload());
        }
      });

      return {
        // Auth
        authKey, authKeyInput, namespace, namespaces, connected,
        saveAuth, disconnectAuth,
        showNsForm, newNsName, createNamespace,
        // Data
        buckets, activeBucket, memories,
        isSearchResult, searchQuery,
        // Pagination
        page, pageSize, total, pageStart, pageEnd, hasPrevPage, hasNextPage,
        nextPage, prevPage, changePageSize,
        // Add memory
        showAddForm, newMemory, saving,
        addMemory,
        // Ingest document
        showIngestForm, ingestFiles, ingestBucket, ingestNewBucket, ingestReplace,
        ingestCreateParent, ingestParentContent,
        ingesting, ingestProgress, ingestDragging,
        ingestAvailable, supportedFormats, acceptAttr, formatsLabel, ingestTotalKb, canIngest,
        onIngestPick, onIngestDrop, ingestDocument,
        // Inline edit
        editingId, editContent, editBucket, editNewBucket, updatingId,
        startEdit, cancelEdit, saveEdit,
        // Connections
        expandedId, newEdge, edgeResults, edgeSearching,
        toggleConnections, addEdge, removeEdge, searchEdgeTargets, pickEdgeTarget,
        // Bucket management
        showBucketForm, newBucketName, renamingBucket, renameBucketValue,
        createBucket, startRenameBucket, confirmRenameBucket, clearBucket,
        // Actions
        loading, searching, error, success,
        reload, onNsChange, loadBuckets, loadMemories,
        searchMemories, clearSearch, deleteMemory,
        selectBucket, formatDate,
      };
    },
  }).mount("#memory-app");
});
